### ISSUE TITLE
Carry arguments through backward pass of custom_vjp

### ISSUE THREAD

**Issue by EddieCunningham** [google/jax#2887]
Is it possible to add a carry argument to the backward pass of a custom_vjp?  It would be useful for writing memory efficient gradients for invertible networks, in particular algorithm 1 of this paper https://pdfs.semanticscholar.org/6c6d/753ff2093452323f1c20ac86f43e707fff6a.pdf.

I was imagining that the code would look something like this:

    import jax
    def low_mem_wrapper(flow):
        x_to_z, z_to_x = flow

        @jax.custom_vjp
        def low_mem_x_to_z(params, log_det_x, x):
            log_det_z, z = x_to_z(params, log_det_x, x)
            return log_det_z, z

        def fwd(params, log_det_x, x):
            return low_mem_x_to_z(params, log_det_x, x), ()

        def bwd(res, g, carry):
            params, log_det_z, z = carry
            log_det_x, x = z_to_x(params, log_det_z, z)
            _, vjp = jax.vjp(x_to_z, params, log_det_x, x)
            return vjp(g), (params, log_det_x, x)

        low_mem_x_to_z.defvjp(fwd, bwd)

        return low_mem_x_to_z, z_to_x

Thanks for the great work in JAX!

**Comment by mattjj** [comment]
Thanks for the question, and the kind words!

Unfortunately I can't think of a nice way to do this. The carry would have to be handled by [the backward pass code](https://github.com/google/jax/blob/3e87e8f9608338b9a44f85a8164e02f46a801968/jax/interpreters/ad.py#L225). You could do it with [your own jaxpr interpreter](https://jax.readthedocs.io/en/latest/notebooks/Writing_custom_interpreters_in_Jax.html) (since `backward_pass` in ad.py is essentially a jaxpr interpreter), and then you could carry along whatever auxiliary variables you want, but that might be diving more into JAX internals than you'd like.

@apaszke has been investigating invertible/reversible sub-computations recently.

By the way, this isn't about your main question but I thought I'd point it out: instead of closing over `x_to_z` and `z_to_x` I think you can also write things like this:

```python
from functools import partial

@partial(jax.custom_vjp, static_argnums=(0, 1))
def low_mem_x_to_z(x_to_z, z_to_x, params, log_det_x, x):
    log_det_z, z = x_to_z(params, log_det_x, x)
    return log_det_z, z

def fwd(x_to_z, z_to_x, params, log_det_x, x):
    return low_mem_x_to_z(params, log_det_x, x), ()

def bwd(x_to_z, z_to_x, res, g, carry):
    params, log_det_z, z = carry
    log_det_x, x = z_to_x(params, log_det_z, z)
    _, vjp = jax.vjp(x_to_z, params, log_det_x, x)
    return vjp(g), (params, log_det_x, x)

low_mem_x_to_z.defvjp(fwd, bwd)
```

**Comment by EddieCunningham** [comment]
I'll take a look at the custom interpreter, thanks!

**Comment by BohdanK-1QBit** [comment]
I've encountered a similar issue. Wondering whether introducing an extra dummy argument to a function can be used to comply with Jax check for the matching shape of the bwd()' output and fwd()'s input. For example as in the following code:
